### ISSUE TITLE
Date values are processed inconsistently in formulas

### ISSUE THREAD

**Issue by Pankraty** [ClosedXML/ClosedXML#729]
## Read and complete the full issue template

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

current develop

**What is the current behavior?**

If cells with date values are included in the formula as parts of the range (```SUM(A1:A2)```) their values are ignored by ClosedXML. But when such cells are referred directly (```A1+A2```) they are treated like numbers. 

Excel processes these cases differently: ```SUM(A1:A2)``` produces a number, ```A1+A2``` returns date.

**What is the expected behavior or new feature?**

The behavior should be either the same in both cases (always treat as numbers) or similar to Excel's. This is debatable.

**Did this work in previous versions of our tool?  Which versions?**

No. Before #696 this failed with NullReferenceException

## Reproducibility
**Code to reproduce problem:**
```c#
public void Main()
{
            using (var wb = new XLWorkbook())
            {
                var ws = wb.AddWorksheet("Test SUM");
                ws.Cell("A1").Value = 1;
                ws.Cell("A2").Value = new DateTime(2018, 1, 1);
                ws.Cell("A3").FormulaA1 = "SUM(A1:A2)";
                ws.Cell("A4").FormulaA1 = "A1+A2";
                Console.WriteLine("Expected: 43102, actual: {0}", ws.Cell("A3").Value);
                Console.WriteLine("Expected: 02.01.2018, actual: {0}", ws.Cell("A4").Value);
            }

            Console.WriteLine("Press any key to continue");
            Console.ReadKey();
}
```
- [x] I attached a sample spreadsheet.  (You can drag files on to this issue)

**Comment by igitur** [comment]
Yes, this is a known issue, and I couldn't find an easy fix. There's a lot of implicit type casting in the calculation engine. My usual recommendation is just to set the data type and number format of the cell explicitly.

**Comment by igitur** [comment]
Note that even Excel gets it wrong sometimes. When you have a cell that has a formula equal to the difference between 2 dates, it defaults to a date type, where one would expect a numeric type (the difference between 2 dates is numeric, not another date).

**Comment by Pankraty** [comment]
The automatic type cast is only one side, and maybe there is no point in emulating Excel's behavior (yes, I remember, we don't create a headless Excel)). Another side is that no matter what typecast is performed Excel returns the same value: in my example, this was either 43102 or 02.01.2018 which are the same. In your example, with date difference, this would be either 365 or 30.12.1900 which are the same again. ClosedXML, on the other hand, returns 1 or 02.01.2018 depending on what formula is used, and those values are not the same of course.

It is not something to be fixed urgently - after all, this caused NullRefs for a long time and no one noticed. I just found it accidentally while preparing unit tests for a completely different task and decided to write it down.

**Comment by igitur** [comment]
Oh! The actual value is 1! Yes, that is strange. I'll look again.
